diff --git a/src/body.rs b/src/body.rs
index 3c3f6a1..700cdc8 100644
--- a/src/body.rs
+++ b/src/body.rs
@@ -13,14 +13,6 @@ use sync_wrapper::SyncWrapper;
 
 type BoxBody = http_body_util::combinators::UnsyncBoxBody<Bytes, Error>;
 
-fn boxed<B>(body: B) -> BoxBody
-where
-    B: http_body::Body<Data = Bytes> + Send + 'static,
-    B::Error: Into<BoxError>,
-{
-    try_downcast(body).unwrap_or_else(|body| body.map_err(Error::new).boxed_unsync())
-}
-
 pub(crate) fn try_downcast<T, K>(k: K) -> Result<T, K>
 where
     T: 'static,
@@ -28,9 +20,9 @@ where
 {
     let mut k = Some(k);
     if let Some(k) = <dyn std::any::Any>::downcast_mut::<Option<T>>(&mut k) {
-        Ok(k.take().unwrap())
+        Ok(unsafe { k.take().unwrap_unchecked() })
     } else {
-        Err(k.unwrap())
+        Err(unsafe { k.unwrap_unchecked() })
     }
 }
 
@@ -45,7 +37,7 @@ impl Body {
         B: http_body::Body<Data = Bytes> + Send + 'static,
         B::Error: Into<BoxError>,
     {
-        try_downcast(body).unwrap_or_else(|body| Self(boxed(body)))
+        try_downcast(body).unwrap_or_else(|body| Self(body.map_err(Error::new).boxed_unsync()))
     }
 
     /// Create an empty body.
diff --git a/src/ext_traits/mod.rs b/src/ext_traits/mod.rs
index 02595fb..04d87f8 100644
--- a/src/ext_traits/mod.rs
+++ b/src/ext_traits/mod.rs
@@ -6,13 +6,11 @@ mod tests {
     use std::convert::Infallible;
 
     use crate::extract::{FromRef, FromRequestParts};
-    use async_trait::async_trait;
     use http::request::Parts;
 
     #[derive(Debug, Default, Clone, Copy)]
     pub(crate) struct State<S>(pub(crate) S);
 
-    #[async_trait]
     impl<OuterState, InnerState> FromRequestParts<OuterState> for State<InnerState>
     where
         InnerState: FromRef<OuterState>,
@@ -32,7 +30,6 @@ mod tests {
     // some extractor that requires the state, such as `SignedCookieJar`
     pub(crate) struct RequiresState(pub(crate) String);
 
-    #[async_trait]
     impl<S> FromRequestParts<S> for RequiresState
     where
         S: Send + Sync,
diff --git a/src/ext_traits/request.rs b/src/ext_traits/request.rs
index 5b7aee7..37080b0 100644
--- a/src/ext_traits/request.rs
+++ b/src/ext_traits/request.rs
@@ -1,6 +1,6 @@
 use crate::body::Body;
 use crate::extract::{DefaultBodyLimitKind, FromRequest, FromRequestParts, Request};
-use futures_util::future::BoxFuture;
+use core::future::Future;
 
 mod sealed {
     pub trait Sealed {}
@@ -20,7 +20,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///
     /// ```
     /// use axum::{
-    ///     async_trait,
     ///     extract::{Request, FromRequest},
     ///     body::Body,
     ///     http::{header::CONTENT_TYPE, StatusCode},
@@ -30,7 +29,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///
     /// struct FormOrJson<T>(T);
     ///
-    /// #[async_trait]
     /// impl<S, T> FromRequest<S> for FormOrJson<T>
     /// where
     ///     Json<T>: FromRequest<()>,
@@ -67,7 +65,7 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///     }
     /// }
     /// ```
-    fn extract<E, M>(self) -> BoxFuture<'static, Result<E, E::Rejection>>
+    fn extract<E, M>(self) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequest<(), M> + 'static,
         M: 'static;
@@ -83,7 +81,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///
     /// ```
     /// use axum::{
-    ///     async_trait,
     ///     body::Body,
     ///     extract::{Request, FromRef, FromRequest},
     ///     RequestExt,
@@ -93,7 +90,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///     requires_state: RequiresState,
     /// }
     ///
-    /// #[async_trait]
     /// impl<S> FromRequest<S> for MyExtractor
     /// where
     ///     String: FromRef<S>,
@@ -111,7 +107,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     /// // some extractor that consumes the request body and requires state
     /// struct RequiresState { /* ... */ }
     ///
-    /// #[async_trait]
     /// impl<S> FromRequest<S> for RequiresState
     /// where
     ///     String: FromRef<S>,
@@ -124,7 +119,7 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///     # }
     /// }
     /// ```
-    fn extract_with_state<E, S, M>(self, state: &S) -> BoxFuture<'_, Result<E, E::Rejection>>
+    fn extract_with_state<E, S, M>(self, state: &S) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequest<S, M> + 'static,
         S: Send + Sync;
@@ -137,7 +132,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///
     /// ```
     /// use axum::{
-    ///     async_trait,
     ///     extract::{Path, Request, FromRequest},
     ///     response::{IntoResponse, Response},
     ///     body::Body,
@@ -154,7 +148,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///     payload: T,
     /// }
     ///
-    /// #[async_trait]
     /// impl<S, T> FromRequest<S> for MyExtractor<T>
     /// where
     ///     S: Send + Sync,
@@ -179,7 +172,7 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///     }
     /// }
     /// ```
-    fn extract_parts<E>(&mut self) -> BoxFuture<'_, Result<E, E::Rejection>>
+    fn extract_parts<E>(&mut self) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequestParts<()> + 'static;
 
@@ -191,7 +184,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///
     /// ```
     /// use axum::{
-    ///     async_trait,
     ///     extract::{Request, FromRef, FromRequest, FromRequestParts},
     ///     http::request::Parts,
     ///     response::{IntoResponse, Response},
@@ -204,7 +196,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///     payload: T,
     /// }
     ///
-    /// #[async_trait]
     /// impl<S, T> FromRequest<S> for MyExtractor<T>
     /// where
     ///     String: FromRef<S>,
@@ -234,7 +225,6 @@ pub trait RequestExt: sealed::Sealed + Sized {
     ///
     /// struct RequiresState {}
     ///
-    /// #[async_trait]
     /// impl<S> FromRequestParts<S> for RequiresState
     /// where
     ///     String: FromRef<S>,
@@ -250,7 +240,7 @@ pub trait RequestExt: sealed::Sealed + Sized {
     fn extract_parts_with_state<'a, E, S>(
         &'a mut self,
         state: &'a S,
-    ) -> BoxFuture<'a, Result<E, E::Rejection>>
+    ) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequestParts<S> + 'static,
         S: Send + Sync;
@@ -267,7 +257,7 @@ pub trait RequestExt: sealed::Sealed + Sized {
 }
 
 impl RequestExt for Request {
-    fn extract<E, M>(self) -> BoxFuture<'static, Result<E, E::Rejection>>
+    fn extract<E, M>(self) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequest<(), M> + 'static,
         M: 'static,
@@ -275,7 +265,7 @@ impl RequestExt for Request {
         self.extract_with_state(&())
     }
 
-    fn extract_with_state<E, S, M>(self, state: &S) -> BoxFuture<'_, Result<E, E::Rejection>>
+    fn extract_with_state<E, S, M>(self, state: &S) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequest<S, M> + 'static,
         S: Send + Sync,
@@ -283,40 +273,38 @@ impl RequestExt for Request {
         E::from_request(self, state)
     }
 
-    fn extract_parts<E>(&mut self) -> BoxFuture<'_, Result<E, E::Rejection>>
+    fn extract_parts<E>(&mut self) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequestParts<()> + 'static,
     {
         self.extract_parts_with_state(&())
     }
 
-    fn extract_parts_with_state<'a, E, S>(
+    async fn extract_parts_with_state<'a, E, S>(
         &'a mut self,
         state: &'a S,
-    ) -> BoxFuture<'a, Result<E, E::Rejection>>
+    ) -> Result<E, E::Rejection>
     where
         E: FromRequestParts<S> + 'static,
         S: Send + Sync,
     {
-        let mut req = Request::new(());
-        *req.version_mut() = self.version();
-        *req.method_mut() = self.method().clone();
-        *req.uri_mut() = self.uri().clone();
-        *req.headers_mut() = std::mem::take(self.headers_mut());
-        *req.extensions_mut() = std::mem::take(self.extensions_mut());
-        let (mut parts, ()) = req.into_parts();
-
-        Box::pin(async move {
-            let result = E::from_request_parts(&mut parts, state).await;
-
-            *self.version_mut() = parts.version;
-            *self.method_mut() = parts.method.clone();
-            *self.uri_mut() = parts.uri.clone();
-            *self.headers_mut() = std::mem::take(&mut parts.headers);
-            *self.extensions_mut() = std::mem::take(&mut parts.extensions);
-
-            result
-        })
+        let mut parts = http::request::Parts {
+            method: std::mem::take(self.method_mut()),
+            uri: std::mem::take(self.uri_mut()),
+            version: self.version(),
+            headers: std::mem::take(self.headers_mut()),
+            extensions: std::mem::take(self.extensions_mut()),
+        }
+
+        let result = E::from_request_parts(&mut parts, state).await;
+
+        *self.version_mut() = parts.version;
+        std::mem::swap(self.method_mut(), &mut parts.method);
+        std::mem::swap(self.uri_mut(), &mut parts.uri);
+        std::mem::swap(self.headers_mut(), &mut parts.headers);
+        std::mem::swap(self.extensions_mut(), &mut parts.extensions);;
+
+        result
     }
 
     fn with_limited_body(self) -> Request {
@@ -345,7 +333,6 @@ mod tests {
         ext_traits::tests::{RequiresState, State},
         extract::FromRef,
     };
-    use async_trait::async_trait;
     use http::Method;
 
     #[tokio::test]
@@ -414,7 +401,6 @@ mod tests {
         body: String,
     }
 
-    #[async_trait]
     impl<S> FromRequest<S> for WorksForCustomExtractor
     where
         S: Send + Sync,
diff --git a/src/ext_traits/request_parts.rs b/src/ext_traits/request_parts.rs
index e7063f4..9362331 100644
--- a/src/ext_traits/request_parts.rs
+++ b/src/ext_traits/request_parts.rs
@@ -1,5 +1,5 @@
 use crate::extract::FromRequestParts;
-use futures_util::future::BoxFuture;
+use core::future::Future;
 use http::request::Parts;
 
 mod sealed {
@@ -21,7 +21,6 @@ pub trait RequestPartsExt: sealed::Sealed + Sized {
     ///     response::{Response, IntoResponse},
     ///     http::request::Parts,
     ///     RequestPartsExt,
-    ///     async_trait,
     /// };
     /// use std::collections::HashMap;
     ///
@@ -30,7 +29,6 @@ pub trait RequestPartsExt: sealed::Sealed + Sized {
     ///     query_params: HashMap<String, String>,
     /// }
     ///
-    /// #[async_trait]
     /// impl<S> FromRequestParts<S> for MyExtractor
     /// where
     ///     S: Send + Sync,
@@ -54,7 +52,7 @@ pub trait RequestPartsExt: sealed::Sealed + Sized {
     ///     }
     /// }
     /// ```
-    fn extract<E>(&mut self) -> BoxFuture<'_, Result<E, E::Rejection>>
+    fn extract<E>(&mut self) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequestParts<()> + 'static;
 
@@ -70,14 +68,12 @@ pub trait RequestPartsExt: sealed::Sealed + Sized {
     ///     response::{Response, IntoResponse},
     ///     http::request::Parts,
     ///     RequestPartsExt,
-    ///     async_trait,
     /// };
     ///
     /// struct MyExtractor {
     ///     requires_state: RequiresState,
     /// }
     ///
-    /// #[async_trait]
     /// impl<S> FromRequestParts<S> for MyExtractor
     /// where
     ///     String: FromRef<S>,
@@ -97,7 +93,6 @@ pub trait RequestPartsExt: sealed::Sealed + Sized {
     /// struct RequiresState { /* ... */ }
     ///
     /// // some extractor that requires a `String` in the state
-    /// #[async_trait]
     /// impl<S> FromRequestParts<S> for RequiresState
     /// where
     ///     String: FromRef<S>,
@@ -113,14 +108,14 @@ pub trait RequestPartsExt: sealed::Sealed + Sized {
     fn extract_with_state<'a, E, S>(
         &'a mut self,
         state: &'a S,
-    ) -> BoxFuture<'a, Result<E, E::Rejection>>
+    ) -> impl Future<Output = Result<E, E::Rejection>> + Send
     where
         E: FromRequestParts<S> + 'static,
         S: Send + Sync;
 }
 
 impl RequestPartsExt for Parts {
-    fn extract<E>(&mut self) -> BoxFuture<'_, Result<E, E::Rejection>>
+    fn extract<E>(&mut self) -> impl Future<Output = Result<E, E::Rejection>>
     where
         E: FromRequestParts<()> + 'static,
     {
@@ -130,7 +125,7 @@ impl RequestPartsExt for Parts {
     fn extract_with_state<'a, E, S>(
         &'a mut self,
         state: &'a S,
-    ) -> BoxFuture<'a, Result<E, E::Rejection>>
+    ) -> impl Future<Output = Result<E, E::Rejection>>
     where
         E: FromRequestParts<S> + 'static,
         S: Send + Sync,
@@ -148,7 +143,6 @@ mod tests {
         ext_traits::tests::{RequiresState, State},
         extract::FromRef,
     };
-    use async_trait::async_trait;
     use http::{Method, Request};
 
     #[tokio::test]
@@ -181,7 +175,6 @@ mod tests {
         from_state: String,
     }
 
-    #[async_trait]
     impl<S> FromRequestParts<S> for WorksForCustomExtractor
     where
         S: Send + Sync,
diff --git a/src/extract/mod.rs b/src/extract/mod.rs
index c8e2d21..f5e3c40 100644
--- a/src/extract/mod.rs
+++ b/src/extract/mod.rs
@@ -5,7 +5,7 @@
 //! [`axum::extract`]: https://docs.rs/axum/0.7/axum/extract/index.html
 
 use crate::{body::Body, response::IntoResponse};
-use async_trait::async_trait;
+use core::future::Future;
 use http::request::Parts;
 use std::convert::Infallible;
 
@@ -42,7 +42,6 @@ mod private {
 /// See [`axum::extract`] for more general docs about extractors.
 ///
 /// [`axum::extract`]: https://docs.rs/axum/0.7/axum/extract/index.html
-#[async_trait]
 #[cfg_attr(
     nightly_error_messages,
     diagnostic::on_unimplemented(
@@ -55,7 +54,7 @@ pub trait FromRequestParts<S>: Sized {
     type Rejection: IntoResponse;
 
     /// Perform the extraction.
-    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection>;
+    fn from_request_parts(parts: &mut Parts, state: &S) -> impl Future<Output = Result<Self, Self::Rejection>> + Send;
 }
 
 /// Types that can be created from requests.
@@ -69,7 +68,6 @@ pub trait FromRequestParts<S>: Sized {
 /// See [`axum::extract`] for more general docs about extractors.
 ///
 /// [`axum::extract`]: https://docs.rs/axum/0.7/axum/extract/index.html
-#[async_trait]
 #[cfg_attr(
     nightly_error_messages,
     diagnostic::on_unimplemented(
@@ -82,10 +80,9 @@ pub trait FromRequest<S, M = private::ViaRequest>: Sized {
     type Rejection: IntoResponse;
 
     /// Perform the extraction.
-    async fn from_request(req: Request, state: &S) -> Result<Self, Self::Rejection>;
+    fn from_request(req: Request, state: &S) -> impl Future<Output = Result<Self, Self::Rejection>> + Send;
 }
 
-#[async_trait]
 impl<S, T> FromRequest<S, private::ViaParts> for T
 where
     S: Send + Sync,
@@ -99,7 +96,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S, T> FromRequestParts<S> for Option<T>
 where
     T: FromRequestParts<S>,
@@ -115,7 +111,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S, T> FromRequest<S> for Option<T>
 where
     T: FromRequest<S>,
@@ -128,7 +123,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S, T> FromRequestParts<S> for Result<T, T::Rejection>
 where
     T: FromRequestParts<S>,
@@ -141,7 +135,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S, T> FromRequest<S> for Result<T, T::Rejection>
 where
     T: FromRequest<S>,
diff --git a/src/extract/request_parts.rs b/src/extract/request_parts.rs
index 73f54db..13d11a9 100644
--- a/src/extract/request_parts.rs
+++ b/src/extract/request_parts.rs
@@ -1,12 +1,10 @@
 use super::{rejection::*, FromRequest, FromRequestParts, Request};
 use crate::{body::Body, RequestExt};
-use async_trait::async_trait;
 use bytes::Bytes;
 use http::{request::Parts, Extensions, HeaderMap, Method, Uri, Version};
 use http_body_util::BodyExt;
 use std::convert::Infallible;
 
-#[async_trait]
 impl<S> FromRequest<S> for Request
 where
     S: Send + Sync,
@@ -18,7 +16,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S> FromRequestParts<S> for Method
 where
     S: Send + Sync,
@@ -30,7 +27,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S> FromRequestParts<S> for Uri
 where
     S: Send + Sync,
@@ -42,7 +38,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S> FromRequestParts<S> for Version
 where
     S: Send + Sync,
@@ -59,7 +54,6 @@ where
 /// Prefer using [`TypedHeader`] to extract only the headers you need.
 ///
 /// [`TypedHeader`]: https://docs.rs/axum/0.7/axum/extract/struct.TypedHeader.html
-#[async_trait]
 impl<S> FromRequestParts<S> for HeaderMap
 where
     S: Send + Sync,
@@ -71,7 +65,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S> FromRequest<S> for Bytes
 where
     S: Send + Sync,
@@ -90,7 +83,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S> FromRequest<S> for String
 where
     S: Send + Sync,
@@ -106,15 +98,10 @@ where
                 }
             })?;
 
-        let string = std::str::from_utf8(&bytes)
-            .map_err(InvalidUtf8::from_err)?
-            .to_owned();
-
-        Ok(string)
+        String::from_utf8(bytes.into()).map_err(|e| InvalidUtf8::from_err(e).into())
     }
 }
 
-#[async_trait]
 impl<S> FromRequestParts<S> for Parts
 where
     S: Send + Sync,
@@ -126,7 +113,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S> FromRequestParts<S> for Extensions
 where
     S: Send + Sync,
@@ -138,7 +124,6 @@ where
     }
 }
 
-#[async_trait]
 impl<S> FromRequest<S> for Body
 where
     S: Send + Sync,
diff --git a/src/extract/tuple.rs b/src/extract/tuple.rs
index 021b961..cbd91a7 100644
--- a/src/extract/tuple.rs
+++ b/src/extract/tuple.rs
@@ -1,10 +1,8 @@
 use super::{FromRequest, FromRequestParts, Request};
 use crate::response::{IntoResponse, Response};
-use async_trait::async_trait;
 use http::request::Parts;
 use std::convert::Infallible;
 
-#[async_trait]
 impl<S> FromRequestParts<S> for ()
 where
     S: Send + Sync,
@@ -20,7 +18,6 @@ macro_rules! impl_from_request {
     (
         [$($ty:ident),*], $last:ident
     ) => {
-        #[async_trait]
         #[allow(non_snake_case, unused_mut, unused_variables)]
         impl<S, $($ty,)* $last> FromRequestParts<S> for ($($ty,)* $last,)
         where
@@ -46,7 +43,6 @@ macro_rules! impl_from_request {
 
         // This impl must not be generic over M, otherwise it would conflict with the blanket
         // implementation of `FromRequest<S, Mut>` for `T: FromRequestParts<S>`.
-        #[async_trait]
         #[allow(non_snake_case, unused_mut, unused_variables)]
         impl<S, $($ty,)* $last> FromRequest<S> for ($($ty,)* $last,)
         where
diff --git a/src/response/into_response.rs b/src/response/into_response.rs
index 15608b1..5e0703c 100644
--- a/src/response/into_response.rs
+++ b/src/response/into_response.rs
@@ -367,7 +367,7 @@ where
 {
     fn into_response(self) -> Response {
         let (parts, res) = self;
-        (parts.status, parts.headers, parts.extensions, res).into_response()
+        (parts.status, parts.version, parts.headers, parts.extensions, res).into_response()
     }
 }
 
diff --git a/src/response/into_response_parts.rs b/src/response/into_response_parts.rs
index 4be4fc5..5ead875 100644
--- a/src/response/into_response_parts.rs
+++ b/src/response/into_response_parts.rs
@@ -1,7 +1,7 @@
 use super::{IntoResponse, Response};
 use http::{
     header::{HeaderMap, HeaderName, HeaderValue},
-    Extensions, StatusCode,
+    Extensions, StatusCode, Version,
 };
 use std::{convert::Infallible, fmt};
 
@@ -104,6 +104,16 @@ pub struct ResponseParts {
 }
 
 impl ResponseParts {
+    /// Gets a reference to the response version.
+    pub fn version(&self) -> Version {
+        self.res.version()
+    }
+
+    /// Gets a mutable reference to the response version.
+    pub fn version_mut(&mut self) -> &mut Version {
+        self.res.version_mut()
+    }
+
     /// Gets a reference to the response headers.
     pub fn headers(&self) -> &HeaderMap {
         self.res.headers()
@@ -259,6 +269,15 @@ impl IntoResponseParts for Extensions {
     }
 }
 
+impl IntoResponseParts for Version {
+    type Error = Infallible;
+
+    fn into_response_parts(self, mut res: ResponseParts) -> Result<ResponseParts, Self::Error> {
+        *res.version_mut() = self;
+        Ok(res)
+    }
+}
+
 impl IntoResponseParts for () {
     type Error = Infallible;
 
