diff --git a/src/extension.rs b/src/extension.rs
index e4d170f..9485443 100644
--- a/src/extension.rs
+++ b/src/extension.rs
@@ -1,5 +1,4 @@
 use crate::{extract::rejection::*, response::IntoResponseParts};
-use async_trait::async_trait;
 use axum_core::{
     extract::FromRequestParts,
     response::{IntoResponse, Response, ResponseParts},
@@ -70,7 +69,6 @@ use tower_service::Service;
 #[must_use]
 pub struct Extension<T>(pub T);
 
-#[async_trait]
 impl<T, S> FromRequestParts<S> for Extension<T>
 where
     T: Clone + Send + Sync + 'static,
diff --git a/src/extract/connect_info.rs b/src/extract/connect_info.rs
index f77db6d..3d8f9a0 100644
--- a/src/extract/connect_info.rs
+++ b/src/extract/connect_info.rs
@@ -7,7 +7,6 @@
 use crate::extension::AddExtension;
 
 use super::{Extension, FromRequestParts};
-use async_trait::async_trait;
 use http::request::Parts;
 use std::{
     convert::Infallible,
@@ -139,7 +138,6 @@ opaque_future! {
 #[derive(Clone, Copy, Debug)]
 pub struct ConnectInfo<T>(pub T);
 
-#[async_trait]
 impl<S, T> FromRequestParts<S> for ConnectInfo<T>
 where
     S: Send + Sync,
diff --git a/src/extract/host.rs b/src/extract/host.rs
index f1d179a..62f0dc7 100644
--- a/src/extract/host.rs
+++ b/src/extract/host.rs
@@ -2,7 +2,6 @@ use super::{
     rejection::{FailedToResolveHost, HostRejection},
     FromRequestParts,
 };
-use async_trait::async_trait;
 use http::{
     header::{HeaderMap, FORWARDED},
     request::Parts,
@@ -23,7 +22,6 @@ const X_FORWARDED_HOST_HEADER_KEY: &str = "X-Forwarded-Host";
 #[derive(Debug, Clone)]
 pub struct Host(pub String);
 
-#[async_trait]
 impl<S> FromRequestParts<S> for Host
 where
     S: Send + Sync,
diff --git a/src/extract/matched_path.rs b/src/extract/matched_path.rs
index 6ac0397..99599a9 100644
--- a/src/extract/matched_path.rs
+++ b/src/extract/matched_path.rs
@@ -1,6 +1,5 @@
 use super::{rejection::*, FromRequestParts};
 use crate::routing::{RouteId, NEST_TAIL_PARAM_CAPTURE};
-use async_trait::async_trait;
 use http::request::Parts;
 use std::{collections::HashMap, sync::Arc};
 
@@ -63,7 +62,6 @@ impl MatchedPath {
     }
 }
 
-#[async_trait]
 impl<S> FromRequestParts<S> for MatchedPath
 where
     S: Send + Sync,
diff --git a/src/extract/multipart.rs b/src/extract/multipart.rs
index 7a303a4..38ebaf9 100644
--- a/src/extract/multipart.rs
+++ b/src/extract/multipart.rs
@@ -4,7 +4,6 @@
 
 use super::{FromRequest, Request};
 use crate::body::Bytes;
-use async_trait::async_trait;
 use axum_core::{
     __composite_rejection as composite_rejection, __define_rejection as define_rejection,
     response::{IntoResponse, Response},
@@ -65,7 +64,6 @@ pub struct Multipart {
     inner: multer::Multipart<'static>,
 }
 
-#[async_trait]
 impl<S> FromRequest<S> for Multipart
 where
     S: Send + Sync,
diff --git a/src/extract/nested_path.rs b/src/extract/nested_path.rs
index 72712a4..61966a0 100644
--- a/src/extract/nested_path.rs
+++ b/src/extract/nested_path.rs
@@ -4,7 +4,6 @@ use std::{
 };
 
 use crate::extract::Request;
-use async_trait::async_trait;
 use axum_core::extract::FromRequestParts;
 use http::request::Parts;
 use tower_layer::{layer_fn, Layer};
@@ -47,7 +46,6 @@ impl NestedPath {
     }
 }
 
-#[async_trait]
 impl<S> FromRequestParts<S> for NestedPath
 where
     S: Send + Sync,
diff --git a/src/extract/path/mod.rs b/src/extract/path/mod.rs
index 330e270..4489843 100644
--- a/src/extract/path/mod.rs
+++ b/src/extract/path/mod.rs
@@ -8,7 +8,6 @@ use crate::{
     routing::url_params::UrlParams,
     util::PercentDecodedStr,
 };
-use async_trait::async_trait;
 use axum_core::response::{IntoResponse, Response};
 use http::{request::Parts, StatusCode};
 use serde::de::DeserializeOwned;
@@ -145,7 +144,6 @@ pub struct Path<T>(pub T);
 
 axum_core::__impl_deref!(Path);
 
-#[async_trait]
 impl<T, S> FromRequestParts<S> for Path<T>
 where
     T: DeserializeOwned + Send,
@@ -446,7 +444,6 @@ impl std::error::Error for FailedToDeserializePathParams {}
 #[derive(Debug)]
 pub struct RawPathParams(Vec<(Arc<str>, PercentDecodedStr)>);
 
-#[async_trait]
 impl<S> FromRequestParts<S> for RawPathParams
 where
     S: Send + Sync,
diff --git a/src/extract/query.rs b/src/extract/query.rs
index a331b68..371612b 100644
--- a/src/extract/query.rs
+++ b/src/extract/query.rs
@@ -1,5 +1,4 @@
 use super::{rejection::*, FromRequestParts};
-use async_trait::async_trait;
 use http::{request::Parts, Uri};
 use serde::de::DeserializeOwned;
 
@@ -51,7 +50,6 @@ use serde::de::DeserializeOwned;
 #[derive(Debug, Clone, Copy, Default)]
 pub struct Query<T>(pub T);
 
-#[async_trait]
 impl<T, S> FromRequestParts<S> for Query<T>
 where
     T: DeserializeOwned,
diff --git a/src/extract/raw_form.rs b/src/extract/raw_form.rs
index a4e0d6c..29cb4c6 100644
--- a/src/extract/raw_form.rs
+++ b/src/extract/raw_form.rs
@@ -1,4 +1,3 @@
-use async_trait::async_trait;
 use axum_core::extract::{FromRequest, Request};
 use bytes::Bytes;
 use http::Method;
@@ -30,7 +29,6 @@ use super::{
 #[derive(Debug)]
 pub struct RawForm(pub Bytes);
 
-#[async_trait]
 impl<S> FromRequest<S> for RawForm
 where
     S: Send + Sync,
diff --git a/src/extract/raw_query.rs b/src/extract/raw_query.rs
index d8c56f8..c792960 100644
--- a/src/extract/raw_query.rs
+++ b/src/extract/raw_query.rs
@@ -1,5 +1,4 @@
 use super::FromRequestParts;
-use async_trait::async_trait;
 use http::request::Parts;
 use std::convert::Infallible;
 
@@ -25,7 +24,6 @@ use std::convert::Infallible;
 #[derive(Debug)]
 pub struct RawQuery(pub Option<String>);
 
-#[async_trait]
 impl<S> FromRequestParts<S> for RawQuery
 where
     S: Send + Sync,
diff --git a/src/extract/request_parts.rs b/src/extract/request_parts.rs
index 55bc340..6ad0c7a 100644
--- a/src/extract/request_parts.rs
+++ b/src/extract/request_parts.rs
@@ -1,5 +1,4 @@
 use super::{Extension, FromRequestParts};
-use async_trait::async_trait;
 use http::{request::Parts, Uri};
 use std::convert::Infallible;
 
@@ -70,7 +69,6 @@ use std::convert::Infallible;
 pub struct OriginalUri(pub Uri);
 
 #[cfg(feature = "original-uri")]
-#[async_trait]
 impl<S> FromRequestParts<S> for OriginalUri
 where
     S: Send + Sync,
diff --git a/src/extract/state.rs b/src/extract/state.rs
index fb401c0..e72c2e1 100644
--- a/src/extract/state.rs
+++ b/src/extract/state.rs
@@ -1,4 +1,3 @@
-use async_trait::async_trait;
 use axum_core::extract::{FromRef, FromRequestParts};
 use http::request::Parts;
 use std::{
@@ -219,13 +218,11 @@ use std::{
 /// ```rust
 /// use axum_core::extract::{FromRequestParts, FromRef};
 /// use http::request::Parts;
-/// use async_trait::async_trait;
 /// use std::convert::Infallible;
 ///
 /// // the extractor your library provides
 /// struct MyLibraryExtractor;
 ///
-/// #[async_trait]
 /// impl<S> FromRequestParts<S> for MyLibraryExtractor
 /// where
 ///     // keep `S` generic but require that it can produce a `MyLibraryState`
@@ -344,7 +341,6 @@ use std::{
 #[derive(Debug, Default, Clone, Copy)]
 pub struct State<S>(pub S);
 
-#[async_trait]
 impl<OuterState, InnerState> FromRequestParts<OuterState> for State<InnerState>
 where
     InnerState: FromRef<OuterState>,
diff --git a/src/extract/ws.rs b/src/extract/ws.rs
index cfd10c5..9059d79 100644
--- a/src/extract/ws.rs
+++ b/src/extract/ws.rs
@@ -93,7 +93,6 @@
 use self::rejection::*;
 use super::FromRequestParts;
 use crate::{body::Bytes, response::Response, Error};
-use async_trait::async_trait;
 use axum_core::body::Body;
 use futures_util::{
     sink::{Sink, SinkExt},
@@ -381,7 +380,6 @@ impl OnFailedUpgrade for DefaultOnFailedUpgrade {
     fn call(self, _error: Error) {}
 }
 
-#[async_trait]
 impl<S> FromRequestParts<S> for WebSocketUpgrade<DefaultOnFailedUpgrade>
 where
     S: Send + Sync,
@@ -464,8 +462,8 @@ impl WebSocket {
     /// Receive another message.
     ///
     /// Returns `None` if the stream has closed.
-    pub async fn recv(&mut self) -> Option<Result<Message, Error>> {
-        self.next().await
+    pub fn recv(&mut self) -> impl Future<Output = Option<Result<Message, Error>>> + Send + '_ {
+        self.next()
     }
 
     /// Send a message.
@@ -485,6 +483,10 @@ impl WebSocket {
     pub fn protocol(&self) -> Option<&HeaderValue> {
         self.protocol.as_ref()
     }
+
+    pub fn into_inner(self) -> (WebSocketStream<TokioIo<hyper::upgrade::Upgraded>>, Option<HeaderValue>) {
+        (self.inner, self.protocol)
+    }
 }
 
 impl Stream for WebSocket {
diff --git a/src/form.rs b/src/form.rs
index 966517a..f754c4c 100644
--- a/src/form.rs
+++ b/src/form.rs
@@ -1,6 +1,5 @@
 use crate::extract::Request;
 use crate::extract::{rejection::*, FromRequest, RawForm};
-use async_trait::async_trait;
 use axum_core::response::{IntoResponse, Response};
 use axum_core::RequestExt;
 use http::header::CONTENT_TYPE;
@@ -72,7 +71,6 @@ use serde::Serialize;
 #[must_use]
 pub struct Form<T>(pub T);
 
-#[async_trait]
 impl<T, S> FromRequest<S> for Form<T>
 where
     T: DeserializeOwned,
diff --git a/src/handler/mod.rs b/src/handler/mod.rs
index 579fdbb..9b03eb6 100644
--- a/src/handler/mod.rs
+++ b/src/handler/mod.rs
@@ -196,10 +196,10 @@ where
     Fut: Future<Output = Res> + Send,
     Res: IntoResponse,
 {
-    type Future = Pin<Box<dyn Future<Output = Response> + Send>>;
+    type Future = impl Future<Output = Response> + Send + 'static;
 
     fn call(self, _req: Request, _state: S) -> Self::Future {
-        Box::pin(async move { self().await.into_response() })
+        async move { self().await.into_response() }
     }
 }
 
@@ -217,10 +217,10 @@ macro_rules! impl_handler {
             $( $ty: FromRequestParts<S> + Send, )*
             $last: FromRequest<S, M> + Send,
         {
-            type Future = Pin<Box<dyn Future<Output = Response> + Send>>;
+            type Future = impl Future<Output = Response> + Send + 'static;
 
             fn call(self, req: Request, state: S) -> Self::Future {
-                Box::pin(async move {
+                async move {
                     let (mut parts, body) = req.into_parts();
                     let state = &state;
 
@@ -241,7 +241,7 @@ macro_rules! impl_handler {
                     let res = self($($ty,)* $last,).await;
 
                     res.into_response()
-                })
+                }
             }
         }
     };
diff --git a/src/json.rs b/src/json.rs
index c443592..2d937d4 100644
--- a/src/json.rs
+++ b/src/json.rs
@@ -1,6 +1,5 @@
 use crate::extract::Request;
 use crate::extract::{rejection::*, FromRequest};
-use async_trait::async_trait;
 use axum_core::response::{IntoResponse, Response};
 use bytes::{BufMut, Bytes, BytesMut};
 use http::{
@@ -92,7 +91,6 @@ use serde::{de::DeserializeOwned, Serialize};
 #[must_use]
 pub struct Json<T>(pub T);
 
-#[async_trait]
 impl<T, S> FromRequest<S> for Json<T>
 where
     T: DeserializeOwned,
diff --git a/src/lib.rs b/src/lib.rs
index 601c14a..7264cdb 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -416,6 +416,7 @@
 #![cfg_attr(docsrs, feature(doc_auto_cfg, doc_cfg))]
 #![cfg_attr(test, allow(clippy::float_cmp))]
 #![cfg_attr(not(test), warn(clippy::print_stdout, clippy::dbg_macro))]
+#![feature(impl_trait_in_assoc_type)]
 
 #[macro_use]
 pub(crate) mod macros;
@@ -442,8 +443,6 @@ pub mod serve;
 #[cfg(test)]
 mod test_helpers;
 
-#[doc(no_inline)]
-pub use async_trait::async_trait;
 #[doc(no_inline)]
 pub use http;
 
diff --git a/src/middleware/from_extractor.rs b/src/middleware/from_extractor.rs
index 63ef858..79003af 100644
--- a/src/middleware/from_extractor.rs
+++ b/src/middleware/from_extractor.rs
@@ -39,12 +39,10 @@ use tower_service::Service;
 ///     Router,
 ///     http::{header, StatusCode, request::Parts},
 /// };
-/// use async_trait::async_trait;
 ///
 /// // An extractor that performs authorization.
 /// struct RequireAuth;
 ///
-/// #[async_trait]
 /// impl<S> FromRequestParts<S> for RequireAuth
 /// where
 ///     S: Send + Sync,
@@ -303,7 +301,7 @@ where
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::{async_trait, handler::Handler, routing::get, test_helpers::*, Router};
+    use crate::{handler::Handler, routing::get, test_helpers::*, Router};
     use axum_core::extract::FromRef;
     use http::{header, request::Parts, StatusCode};
     use tower_http::limit::RequestBodyLimitLayer;
@@ -315,7 +313,6 @@ mod tests {
 
         struct RequireAuth;
 
-        #[async_trait::async_trait]
         impl<S> FromRequestParts<S> for RequireAuth
         where
             S: Send + Sync,
@@ -367,7 +364,6 @@ mod tests {
     fn works_with_request_body_limit() {
         struct MyExtractor;
 
-        #[async_trait]
         impl<S> FromRequestParts<S> for MyExtractor
         where
             S: Send + Sync,
