diff --git a/src/date.rs b/src/date.rs
index 8bc0a3b..909a1d9 100644
--- a/src/date.rs
+++ b/src/date.rs
@@ -14,34 +14,19 @@ use crate::Error;
 #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
 pub struct HttpDate {
     /// 0...59
-    sec: u8,
+    pub sec: u8,
     /// 0...59
-    min: u8,
+    pub min: u8,
     /// 0...23
-    hour: u8,
+    pub hour: u8,
     /// 1...31
-    day: u8,
+    pub day: u8,
     /// 1...12
-    mon: u8,
+    pub mon: u8,
     /// 1970...9999
-    year: u16,
+    pub year: u16,
     /// 1...7
-    wday: u8,
-}
-
-impl HttpDate {
-    fn is_valid(&self) -> bool {
-        self.sec < 60
-            && self.min < 60
-            && self.hour < 24
-            && self.day > 0
-            && self.day < 32
-            && self.mon > 0
-            && self.mon <= 12
-            && self.year >= 1970
-            && self.year <= 9999
-            && &HttpDate::from(SystemTime::from(*self)) == self
-    }
+    pub wday: u8,
 }

 impl From<SystemTime> for HttpDate {
@@ -169,9 +154,6 @@ impl FromStr for HttpDate {
         let date = parse_imf_fixdate(x)
             .or_else(|_| parse_rfc850_date(x))
             .or_else(|_| parse_asctime(x))?;
-        if !date.is_valid() {
-            return Err(Error(()));
-        }
         Ok(date)
     }
 }
